### DISCUSSION TITLE
[A-L] (27 marzo 2024) Esercizio 5

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#29] [Esercitazioni A-L]
Sia $G = (V, E)$ un grafo diretto, un vertice $u \in V$ è detto principale se per ogni vertice $v \in V$ esiste un cammino diretto da
$u$ a $v$. Fornire un algoritmo in pseudo-codice che, dato un grafo diretto $G$, determina tutti i vertici principali di $G$. È possibile che tale algoritmo abbia complessità $O(|V|+ |E|)$?

**Comment by cornflexxx** [comment]
### Idea algoritmo:

Per trovare un vertice principale dobbiamo iniziare una $DFS$ da un nodo qualunque $v$, chiamiamo $cc_v$ il numero di nodi trovati durante la $DFS$ a partire da $v$, se $cc_v=|V(G)|$ allora $v$ è un vertice principale. Per trovare ogni nodo principale dobbiamo nel peggiore dei casi fare una $DFS$ su ogni nodo e quindi avremo una complessità $O(n(n+m))$. In realtà però ci basta conoscere un solo $v.P.$ per conoscere poi tutti gli altri. Infatti se supponiamo che $v_1$ sia il vertice principale , allora ogni altro vertice principale $v_i$ formerà un ciclo con $v_1$. Quindi risulta facile trovare tutti gli altri vertici principali una volta che ne abbiamo trovato uno: ci basta visitare ricorsivamente tutti gli archi entranti in $v_1$ e ogni arco su cui passeremo sarà un $v.P.$ (riassumendo una $DFS$ che passa solo per gli archi uscenti).

> [!NOTE]
> Da finire

**Comment by LeoCosta02** [comment]
```C
//Sia G un grafo diretto, un vertice u è detto principale se per ogni vertice v esiste un cammino diretto da u a v 
 //Fornire un algoritmo in pseudo-codice che, dato un grafo diretto determina tutti i vertici principali 
 INPUT:Grafo diretto G=(V,E)
 OUTPUT:insieme A che contiene tutti i vertici principali di G
 PRINCIPAL(G){
    comp[n]=func_globale_scc// vettore contenente le componenti 
    indeg: array dei gradi entranti di GSCC,con ogni coordinata inizializzata a 0 
    A: insieme inizialmente vuoto 
    a=0// mi serve come controllo per vedere se ho due nodi senza archi entranti 
     for each (u appartenente a G){
        for each(v appartenente Adj(u))
         indeg[v]=indeg[v]+1 // in questo array avrò per ogni nodo del grafo il numero di archi entranti 
     }
      for each (u appartenente a G ){
         if(indeg[u]==0 and a!=0) // ho almeno due nodi con grado entrante 0 --> non ho nodi principali 
          return A
          if(indeg[u]==0 and a ==0) // ho almeno un nodo con grado entrante 0 
           a=1
        }
       for each (u appartenente a G  and comp[u]!=0){
         if(indeg[u]==0 and degout(u)!=0) // se ho un solo nodo con grado entrante 0 ma con grado uscente !=0 allora sarà l'unico vertice principale 
          return u 
          else{
            if(indeg[u]!=0 and degout(u)==0) // se ho un  nodo con grado entrante NON nullo e grado uscente !=0 allora esso sicuramente non sarà un nodo principale 
            comp[u]=0 // vuol dire che quel nodo l'ho già esaminato 
            continue // vado avanti con il for each 
             else{
                if(indeg[u]!=0 and degout(u)!=0){ // se ho un nodo con grado entrante ed uscente !=0 allora sono in una comp fort conn  
                    A.add(u)
                    comp[u]=0
                    for(int i=0;i<n;i++){
                        if(comp[u]==comp[i])
                        A.add(i)
                        comp[i]=0
                    }
                }
             }
          }
          
          
       }
        return A 
 }
```

**Reply by LeoCosta02** [reply]
L'algoritmo dovrebbe essere corretto e l'esecuzione nel caso peggiore dovrebbe essere sempre O(N+M).
Penso di aver esaminato tutti i casi possibili per cui non possa esistere un vertice principale o per cui un vertice non può essere principale e il caso in cui abbiamo un solo vertice principale.L'ultimo if esamina il caso in cui ho una componente fortemente connessa che è collegata a tutti i restanti nodi del grafo

**Reply by LeoCosta02** [reply]
L'unica cosa che non ho trovato un modo per migliorare il ciclo for all'interno dell'ultimo if, per eseguire meno cicli possibili ho aggiunto una condizione al for each

**Comment by Frisbon** [comment]
### Soluzione in $O(n+m)$ con Spiegazione
Tradotto dal Wollanese: $x$ **principale sse tutti gli altri nodi lo possono raggiungere.**

Osservazioni da fare:

1. Se tutti i nodi hanno sia archi uscenti che entranti ⇒ tutti i nodi sono principali.
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/efa92707-8092-4257-94e1-c2e092374f79)
2. Se esiste un nodo senza archi entranti ⇒ Sicuramente non è principale
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/5a8af5de-be27-412c-b30f-05862ccef033)
3. Se esiste un nodo senza archi uscenti ⇒ Tutti gli altri non sono principali $\wedge$ bisogna controllare se questo è principale.
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/bfcddd53-7e04-4a6a-af21-22143531e989)
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/f9efce7c-7bfb-401b-98c0-9ad93a3b8f3b)

4. Se esistono DUE nodi senza archi uscenti ⇒ Non ci sono principali.
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/1e734c48-b078-4788-a9a9-0467b6fa19a1)

```
def es5(G):
	caso1 = True
	unico = NULL;
	for nodo in V(G):
		if nodo.inp is empty:
			not_p.add(nodo)
			caso1 = False
		if nodo.out is empty:
			if unico != NULL: #se avevo già trovato un nodo senza archi usc
                return empty #(caso4)
            else:
                caso1 = False
                unico = nodo
	if caso1: return V(G) # ez
	if unico != NULL: #controllo se questo nodo è l'unico principale
		vis = [0...0] ; vis[x] = 1;
		S = stack con x
		while S not empty:
			y = S.top()
			if y.inp not empty: #faccio dfs al contrario lol 
				z = y.inp[0]
				if vis[z] == 0:
					S.push(z)
					vis[z] = 1
			if y.inp is empty:
				S.pop()
		for i in vis: #se tutti so visitati, allora quel nodo è principale
			if i == 0: return empty
			else: return unico					
```

> Il primo for ha costo $O(n)$.
Il while è un DFS (illegale), quindi $O(n+m)$.
L'ultimo for scorre VIS quindi $O(n)$.
Costo finale $O(n+m)$ :)
